labs_api/insights.py

```diff
@@ -0,0 +1,49 @@
+from prediction_market_agent_tooling.markets.omen.omen_subgraph_handler import (
+    HexAddress,
+    OmenSubgraphHandler,
+)
+from prediction_market_agent_tooling.tools.utils import utcnow
+from tavily import TavilyClient
+
+from labs_api.config import Config
+from labs_api.insights_cache import MarketInsightsResponseCache
+from labs_api.models import MarketInsightsResponse, TavilyResponse
+
+
+def market_insights_cached(
+    market_id: HexAddress, cache: MarketInsightsResponseCache
+) -> MarketInsightsResponse:
+    """Returns `market_insights`, but cached daily."""
+    if (cached := cache.find(market_id)) is not None:
+        return cached
```

Comment:
I was wondering if this package [fastapi-cache](https://github.com/long2ice/fastapi-cache) would make sense here?
You could have a logic for caching markets for 3 days (else you fetch a new one, save in the DB and serve that).
This has the advantage of preventing DB lookups for every request.

Reply:
Theoretically, but they say `with backends supporting Redis, Memcached, and Amazon DynamoDB`, we don't have either of them. I was thinking about using Redis, but then I didn't want to complicate it.

> This has the advantage of preventing DB lookups for every request.

I don't follow, how do you prevent lookups for new requests? For each one we need to check if we have a recent summary or not.

Reply:
> Theoretically, but they say with backends supporting Redis, Memcached, and Amazon DynamoDB, we don't have either of them. I was thinking about using Redis, but then I didn't want to complicate it.

They also support `in-memory`, which should be good enough ([link](https://github.com/long2ice/fastapi-cache?tab=readme-ov-file#features)).

> I don't follow, how do you prevent lookups for new requests? For each one we need to check if we have a recent summary or not.

What I meant was - assuming you have in-memory cache, checking an entry by key (e.g. market_id) is much cheaper than the flight time app <> DB, which should only happen if the cache entry is invalid (for ex, after 3 days, we should update the market in the DB). Does that make sense?

Reply:
I see; for in-memory, you don't even need fastapi-cache, there is a built-in decorator that works nicely: https://docs.python.org/3/library/functools.html#functools.cache

However, the time of checking DB for a hit is negligible in comparison to calling Tavily/LLM, and having an in-memory cache on the app-level isn't a good idea.

One problem is growing memory and the second is, the fact that there isn't in-memory cache hit doesn't mean that there won't be in-DB cache hit (for example due to pod randomly restarting, or because of a new release restart, or theoretically we could have multiple pods to handle traffic in the future), so checking DB would be required anyway.